unity-renderer/Assets/DCLPlugins/ECS7/Systems/RaycastSystem/ECSRaycastSystem.cs

```diff
@@ -0,0 +1,228 @@
+using DCL.Configuration;
+using DCL.Controllers;
+using DCL.ECS7;
+using DCL.ECS7.InternalComponents;
+using DCL.ECSComponents;
+using DCL.ECSRuntime;
+using DCL.Helpers;
+using DCL.Models;
+using System.Collections.Generic;
+using UnityEngine;
+using Ray = UnityEngine.Ray;
+using RaycastHit = UnityEngine.RaycastHit;
+using Vector3 = Decentraland.Common.Vector3;
+
+namespace ECSSystems.ECSRaycastSystem
+{
+    public class ECSRaycastSystem
+    {
+        private readonly ECSComponent<PBRaycast> raycastComponent;
+        private readonly ECSComponent<PBRaycastResult> raycastResultComponent;
+        private readonly ECSComponent<PBMeshCollider> meshCollider;
+        private readonly IInternalECSComponent<InternalColliders> physicsColliderComponent;
+        private readonly IInternalECSComponent<InternalColliders> onPointerColliderComponent;
+        private readonly IInternalECSComponent<InternalColliders> customLayerColliderComponent;
+        private readonly IECSComponentWriter componentWriter;
+
+        public ECSRaycastSystem(
+            ECSComponent<PBRaycast> raycastComponent,
+            ECSComponent<PBRaycastResult> raycastResultComponent,
+            ECSComponent<PBMeshCollider> meshCollider,
+            IInternalECSComponent<InternalColliders> physicsColliderComponent,
+            IInternalECSComponent<InternalColliders> onPointerColliderComponent,
+            IInternalECSComponent<InternalColliders> customLayerColliderComponent,
+            IECSComponentWriter componentWriter)
+        {
+            this.raycastComponent = raycastComponent;
+            this.raycastResultComponent = raycastResultComponent;
+            this.meshCollider = meshCollider;
+            this.physicsColliderComponent = physicsColliderComponent;
+            this.onPointerColliderComponent = onPointerColliderComponent;
+            this.customLayerColliderComponent = customLayerColliderComponent;
+            this.componentWriter = componentWriter;
+        }
+
+        public void Update()
+        {
+            var raycasts = raycastComponent.Get();
+            int raycastsCount = raycasts.Count;
+            for (int i = 0; i < raycastsCount; i++)
+            {
+                PBRaycast model = raycasts[i].value.model;
+                IDCLEntity entity = raycasts[i].value.entity;
+                IParcelScene scene = raycasts[i].value.scene;
+
+                if (model.QueryType == RaycastQueryType.RqtNone) continue;
+
+                // If the entity has a raycastResult, the first ray was already cast
+                if (raycastResultComponent.HasComponent(scene, entity))
+                {
+                    bool isContinuous = model.HasContinuous && model.Continuous;
+                    if (!isContinuous && raycastResultComponent.Get(scene, entity).model.Timestamp == model.Timestamp)
+                        continue;
+                }
```

Comment:
maybe we can add an internal component for raycast, create it or update it on raycast component update and removed after being processed (if it is not continuous)
that way we avoid always checking for this

Reply:
Done, check [this commit](https://github.com/decentraland/unity-renderer/pull/4761/commits/08ae36fd041fa1d19a932a19bd9f14a24bd5dd10)